src/transformers/modeling_flash_attention_utils.py

```diff
@@ -220,6 +229,16 @@ def _flash_attention_forward(
         deterministic (`bool`, *optional*):
             Determines if the deterministic option introduced in flash_attn>=2.4.1 is enabled.
     """
+    if is_deepspeed_sp_enabled():
+        spg = ds_comm_groups._get_sequence_parallel_group()
+        # qkv tensors are of shape (batch_size, seq_len, num_heads, head_dim)
+        scatter_idx = 2  # Scatter on num_heads dimension
+        gather_idx = 1  # Gather on seq_len dimension
+        batch_dim_idx = 0  # Synonymous with the batch_first==true
+        query_states = _SeqAllToAll.apply(spg, query_states, scatter_idx, gather_idx, batch_dim_idx)
+        key_states = _SeqAllToAll.apply(spg, key_states, scatter_idx, gather_idx, batch_dim_idx)
+        value_states = _SeqAllToAll.apply(spg, value_states, scatter_idx, gather_idx, batch_dim_idx)
```

Comment:
I am a tad bit unconvinced by checking this for each model that runs flash attention. 
We could: 
- either have self.use_deep_speed, and check for that instead
- separate the deepspeed code, by change the forward if deepspeed is available with deepspeed forward. 
For such changes it is important to make sure we don't introduce regressions in the perfromances. For a single forward pass it's gonna small, but for generation, it might add up!

Reply:
@ArthurZucker, as to your first bullet point, included in `is_deepspeed_sp_enabled` is the check for "use_deep_speed". Generally speaking, I see your point of view but am also concern about duplicated code (your second bullet point).  Is bullet point 1 without burden on user to set another `use_deep_speed` flag sufficient for you? From deepspeed persepective, we assume that if the user enables deepspeed and sets `sp_size > 1` during initialization, it means they intend to use `deepspeed_sp`.

Reply:
>  if the user enables deepspeed and sets sp_size > 1 during initialization, it means they intend to use deepspeed_sp 

yeah, IMO this is enough!

Reply:
WDYT?

Reply:
@ArthurZucker, apologies for late response, was OOF! Your suggestion LGTM, please feel to update this PR (or refactor as a follow up PR).